File: packages/sdk-ts/src/session.ts
import type { SessionExecResponse } from '@sandchest/contract'
import type { HttpClient } from './http.js'
import type { ExecResult, SessionExecOptions } from './types.js'

/**
 * A stateful session inside a sandbox.
 * Sessions persist shell state (working directory, env vars) between commands.
 */
export class Session {
  readonly id: string
  /** @internal */
  readonly _sandboxId: string
  /** @internal */
  readonly _http: HttpClient

  /** @internal — Use `sandbox.session.create()` instead. */
  constructor(id: string, sandboxId: string, http: HttpClient) {
    this.id = id
    this._sandboxId = sandboxId
    this._http = http
  }

  /** Execute a command in this session. State persists between calls. */
  async exec(cmd: string, options?: SessionExecOptions): Promise<ExecResult> {
    const res = await this._http.request<SessionExecResponse>({
      method: 'POST',
      path: `/v1/sandboxes/${this._sandboxId}/sessions/${this.id}/exec`,
      body: {
        cmd,
        timeout_seconds: options?.timeout,
        wait: true,
      },
    })

    return {
      execId: res.exec_id,
      exitCode: res.exit_code,
      stdout: res.stdout,
      stderr: res.stderr,
      durationMs: res.duration_ms,
    }
  }

  /** Destroy this session. */
  async destroy(): Promise<void> {
    await this._http.request<{ ok: true }>({
      method: 'DELETE',
      path: `/v1/sandboxes/${this._sandboxId}/sessions/${this.id}`,
    })
  }
}

File: packages/sdk-py/sandchest/session.py
"""Session — a stateful shell session inside a sandbox."""

from __future__ import annotations

from typing import TYPE_CHECKING

from .types import ExecResult

if TYPE_CHECKING:
    from .http import HttpClient


class Session:
    """A stateful session inside a sandbox.

    Sessions persist shell state (working directory, env vars) between commands.
    """

    def __init__(self, id: str, sandbox_id: str, http: HttpClient) -> None:
        self.id = id
        self._sandbox_id = sandbox_id
        self._http = http

    def exec(
        self,
        cmd: str,
        *,
        timeout: int | None = None,
    ) -> ExecResult:
        """Execute a command in this session. State persists between calls."""
        res = self._http.request(
            "POST",
            f"/v1/sandboxes/{self._sandbox_id}/sessions/{self.id}/exec",
            body={
                "cmd": cmd,
                "timeout_seconds": timeout,
                "wait": True,
            },
        )
        return ExecResult(
            exec_id=res["exec_id"],
            exit_code=res["exit_code"],
            stdout=res["stdout"],
            stderr=res["stderr"],
            duration_ms=res["duration_ms"],
        )

    def destroy(self) -> None:
        """Destroy this session."""
        self._http.request(
            "DELETE",
            f"/v1/sandboxes/{self._sandbox_id}/sessions/{self.id}",
        )
